synapse/storage/databases/main/push_rule.py

```diff
@@ -358,17 +376,17 @@ def get_all_push_rule_updates_txn(txn):
 class PushRuleStore(PushRulesWorkerStore):
     async def add_push_rule(
         self,
-        user_id,
-        rule_id,
-        priority_class,
-        conditions,
-        actions,
-        before=None,
-        after=None,
+        user_id: str,
+        rule_id: str,
+        priority_class: int,
+        conditions: List[Dict[str, str]],
+        actions: List[Union[dict, str]],
+        before: Optional[str] = None,
+        after: Optional[str] = None,
     ) -> None:
         conditions_json = json_encoder.encode(conditions)
         actions_json = json_encoder.encode(actions)
-        async with self._push_rules_stream_id_gen.get_next() as stream_id:
+        async with self._push_rules_stream_id_gen.get_next() as stream_id:  # type: ignore[attr-defined]
```

Comment:
The complaint here is that `PushRulesWorkerStore` declares this id_generator as an `AbstractStreamIdTracker`, which doesn't have the `gen_next` attribute. On non-worker mode, it's actually a `StreamIdGenerator` which does have this method.

Could you try adding this right at the top of the class body? (Before any methods)

```python
    _push_rules_stream_id_gen: AbstractStreamIdGenerator
```

That might make the type-ignore unnecessary.

(cc @squahtx: you might be interested in this)

Reply:
Additional top of this?
https://github.com/matrix-org/synapse/blob/fa3112daef22266bf8a4a37c13a518c29b76eb54/synapse/storage/databases/main/push_rule.py#L117-L132

Reply:
Sorry, at the top of `PushRuleStore`. (It would be incorrect at the top of `PushRuleWorkerStore` because`

Something like this:

```python
class PushRuleStore(PushRulesWorkerStore):
    # Because we have write access, this will be a StreamIdGenerator
    # (see PushRulesWorkerStore.__init__)
    _push_rules_stream_id_gen: AbstractStreamIdGenerator
```

(might need formatting and imports to be resolved).

I'm not sure if this will actually work to make the type-ignore unnecessary though; we can drop this if it doesn't work.

Reply:
It works.

Reply:
Great! (phew!)